diff --git a/bin/nvcc_wrapper b/bin/nvcc_wrapper
index 8a23d0d6..5d8d45df 100755
--- a/bin/nvcc_wrapper
+++ b/bin/nvcc_wrapper
@@ -99,6 +99,9 @@ depfile_target_arg=""
 # Option to remove duplicate libraries and object files
 remove_duplicate_link_files=0
 
+relocatable_set=0
+arch_set2=0
+
 function warn_std_flag() {
   echo "nvcc_wrapper - *warning* you have set multiple standard flags (-std=c++1* or --std=c++1*), only the last is used because nvcc can only accept a single std setting"
 }
@@ -179,9 +182,15 @@ do
     shift
     ;;
   #Handle known nvcc args
-  --dryrun|--verbose|--keep|--keep-dir*|-G|--relocatable-device-code*|-lineinfo|-expt-extended-lambda|--resource-usage|-Xptxas*|--fmad*)
+  --dryrun|--verbose|--keep|--keep-dir*|-G|-lineinfo|-expt-extended-lambda|--resource-usage|-Xptxas*|--fmad*)
     cuda_args="$cuda_args $1"
     ;;
+  --relocatable-device-code**)
+    if [ "x$relocatable_set" == "x0" ]; then
+        cuda_args="$cuda_args $1"
+        relocatable_set=1
+    fi
+    ;;
   #Handle more known nvcc args
   --expt-extended-lambda|--expt-relaxed-constexpr)
     cuda_args="$cuda_args $1"
@@ -275,10 +284,17 @@ do
     ;;
 
   #Handle -arch argument (if its not set use a default) this is the version with = sign
-  -arch*|-gencode*)
+  -gencode*)
     cuda_args="$cuda_args $1"
     arch_set=1
     ;;
+  -arch*)
+    if [ "x$arch_set2" == "x0" ]; then
+        cuda_args="$cuda_args $1"
+        arch_set2=1
+    fi
+    arch_set=1
+    ;;
   #Handle -code argument (if its not set use a default) this is the version with = sign
   -code*)
     cuda_args="$cuda_args $1"
